presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ScalarQueryUtil.java

```diff
@@ -14,29 +14,69 @@
 package com.facebook.presto.sql.planner.optimizations;
 
 import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.ApplyNode;
 import com.facebook.presto.sql.planner.plan.EnforceSingleRowNode;
 import com.facebook.presto.sql.planner.plan.ExchangeNode;
 import com.facebook.presto.sql.planner.plan.FilterNode;
 import com.facebook.presto.sql.planner.plan.PlanNode;
 import com.facebook.presto.sql.planner.plan.PlanVisitor;
 import com.facebook.presto.sql.planner.plan.ProjectNode;
 import com.facebook.presto.sql.planner.plan.ValuesNode;
+import com.facebook.presto.sql.tree.SymbolReference;
 import com.google.common.collect.ImmutableList;
 
+import java.util.function.Function;
+
 import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.function.Function.identity;
 
 public final class ScalarQueryUtil
 {
     private ScalarQueryUtil() {}
 
+    /**
+     * @return true when subquery is scalar and its output symbols are directly mapped to the ApplyNode's output symbols
+     */
+    public static boolean isResolvedScalarSubquery(ApplyNode applyNode)
+    {
+        return isScalar(applyNode.getSubquery(), identity()) && applyNode.getSubqueryAssignments().getExpressions().stream()
+                .allMatch(expression -> expression instanceof SymbolReference);
+    }
+
+    public static boolean isResolvedScalarSubquery(ApplyNode applyNode, Function<PlanNode, PlanNode> lookup)
+    {
+        PlanNode subquery = lookup.apply(applyNode.getSubquery());
+        return isScalar(subquery, lookup) && applyNode.getSubqueryAssignments().getExpressions().stream()
+                .allMatch(expression -> expression instanceof SymbolReference);
+    }
+
     public static boolean isScalar(PlanNode node)
     {
-        return node.accept(new IsScalarPlanVisitor(), null);
+        return isScalar(node, identity());
+    }
+
+    public static boolean isScalar(PlanNode node, Function<PlanNode, PlanNode> lookupFunction)
+    {
+        return node.accept(new IsScalarPlanVisitor(lookupFunction), null);
     }
 
     private static final class IsScalarPlanVisitor
             extends PlanVisitor<Void, Boolean>
     {
+        /*
+        With the iterative optimizer, plan nodes are replaced with
+        GroupReference nodes. This requires the rules that look deeper
+        in the tree than the rewritten node and its immediate sources
+        to use Lookup for resolving the nodes corresponding to the
+        GroupReferences.
+        */
+        private final Function<PlanNode, PlanNode> lookupFunction;
```

Comment:
I'd like to avoid depending on Lookup itself, but the line `return isScalar(node, identity());` several lines above could then be replaced with `return isScalar(node, noLookup()::lookup);` to make things more clear :D